Reduce start delays when using EKF3 and fix time delay bug (Extension of #6290)

WickedShell opened the pull request:
This is only intended to present the last 2 commits for @priseborough (and others) to comment upon which can then be squashed/incorporated into his PR. The changes are to:
A) Allow a backend to report if it is confident in it's estimate of the lag value, which allows us to use it's reported lag immediately, rather then waiting for it to be reported as fully configured. (Significant time savings on a u-blox GPS rather then waiting for the full configuration checks to pass).
B) Rather then assume a rate of 1 measurement interval (which wasn't being correctly bounds checked anyways) we store the worst possible case for any driver that could be loaded, and return that instead. I looked at making it a per GPS driver static method, but it looked way more annoying then a single #define that just had to be maintained. (Since every backend *except* ublox currently uses the same default delay). The effect of this is that the EKF may create a slightly larger then needed buffer for storing the GPS data, but it will at least, A) init, B) doesn't risk having too small a buffer if a u-blox 5 or 6 is later connected.

WickedShell commented:
Please note that there is a behavior change here, the EKF will no longer start if you have set GPS_TYPE to a specific model and that instance hasn't been loaded yet. Because this isn't the default parameter value, and a user has to set this @priseborough and I thought this was acceptable. Also serves to enforce that your backup GPS is working before taking off, which is currently entirely unenforced.

amilcarlucas commented:
Needs a rebase :(

priseborough commented:
@WickedShell Your last commit (as you predicted here https://github.com/ArduPilot/ardupilot/pull/6319) is now breaking it. Can you please fix. TXS.

WickedShell commented:
Rebased.

OXINARF commented on libraries/AP_GPS/AP_GPS_UBLOX.cpp:
```diff
@@ -1307,19 +1307,26 @@ AP_GPS_UBLOX::broadcast_configuration_failure_reason(void) const {
     switch (_hardware_generation) {
+    case UBLOX_UNKNOWN_HARDWARE_GENERATION:
+        lag_sec = 0.22f;
+        // always bail out in this case, it's used to indicate we have yet to recieve a valid
```
recieve -> receive

OXINARF commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -1054,15 +1056,17 @@ bool AP_GPS::get_lag(uint8_t instance, float &lag_sec) const
+        } else {
+            lag_sec = GPS_WORST_LAG_SEC;
+        }
+        return _type[instance] != GPS_TYPE_AUTO;
```
I'm not sure I understand this line. Does this mean that when the user specifies a GPS type but it hasn't been detected yet, it will return the worst lag as valid, but when auto is selected it returns as not valid? It seems the contrary to what I would expect - when a GPS type is selected a GPS is expected to be there so a driver is expected to be created at some point, but with auto there may or may not exist a GPS.

WickedShell replied:
Right so what this is aiming to do is state that:

If we have specified a GPS_TYPE (IE u-blox) then we will wait until that driver has been detected and loaded, and then set the optimal time. However as you observed auto detected instances may never show up. If I never return true from this function then the EKF will never boot. It's accepted that specifying a type is blocking on EKF booting till it's present, but auto will continue to boot. This is the change in behavior mentioned on the dev call.

OXINARF replied:
Ok, that's good, except I'm seeing the opposite. Let me grab your sentence and change it to what I see:

*If we have **not** specified a GPS_TYPE (IE u-blox) then we will wait until **the correct** driver has been detected and loaded, and then set the optimal time.*

_If we have specified a GPS_TYPE (IE u-blox) then we will **set it to worst possible lag and return as VALID immediately.**_

Was that clear?

WickedShell replied:
GAH! Sorry, your absolutely right, rather then ```return _type[instance] != GPS_TYPE_AUTO;``` it should be ```return _type[instance] == GPS_TYPE_AUTO;```

Fixed

OXINARF commented on libraries/AP_GPS/AP_GPS_UBLOX.cpp:
```diff
@@ -1307,19 +1307,26 @@ AP_GPS_UBLOX::broadcast_configuration_failure_reason(void) const {
+        lag_sec = 0.22f;
+        // always bail out in this case, it's used to indicate we have yet to recieve a valid
+        // hardware generation
+        return false;
```
I'm not quite sure this is the correct approach. How does this work with auto config turned off?

WickedShell replied:
Good point, addressed.

OXINARF commented on libraries/AP_GPS/AP_GPS_UBLOX.cpp:
```diff
@@ -1307,19 +1307,26 @@ AP_GPS_UBLOX::broadcast_configuration_failure_reason(void) const {
+        lag_sec = 0.12f;
         break;
     };
+    return !(_unconfigured_messages & CONFIG_VERSION);
```
Why depend on this? If we have the hardware generation, then why not always true? And once again, I'm not sure how this interacts with auto config turned off.

I'm also a bit confused on how we manage the version, we seem to request it by two different ways: MSG_NAV_SVINFO and MSG_MON_VER. But we do different things for both. Specially when receiving MSG_MON_VER we don't remove the CONFIG_VERSION from the unconfigured messages, we wait for it to be requested again to remove it.
My feeling is that we should always request the version instead of being a step of configuration.

WickedShell replied:
MON_VER isn't the same. If you look at what we get back from that, we get a 10 character long string, which while we could build a parse table, is less useful then a enumerated return type which is what we are getting out of SVINFO.

Regarding the configure bit, I agree with you, but we also have 2 ways of tracking if hardware generation is present (we can either check the config bitmask, or check the _hardware_generation parameter), and I went with the bitmask. I think a pass is needed to reduce this scope, however this is entirely outside this PR and I've addressed it.

OXINARF commented on libraries/AP_GPS/GPS_Backend.h:
```diff
@@ -53,8 +53,8 @@ class AP_GPS_Backend
-    // driver specific lag
-    virtual float get_lag(void) const { return 0.2f; }
+    // driver specific lag, returns true if the driver is confident in the provided lag
+    virtual bool get_lag(float &lag) const { lag = 0.2f; return true; }
```
Should this be GPS_WORST_LAG_SEC? This is going to be used for any GPS driver that doesn't have a specific lag, I'm not sure if for those we should assume the worse or not.

WickedShell replied:
This should assume the normal/more expected lag term. The rule of thumb is that the lag is somewhere around one measurement interval, which makes 0.2 a reasonable value as we are expecting them to be using. This is used for all the drivers so changing it would be a behavior change (and anytime a driver is added which is worse would change every other drivers).

OXINARF replied:
OK, I'm convinced on this. Now go convince me of the others :smile:

WickedShell replied:
I happen to agree with you on the others (well and there is one I need to look at context wise to see who's right... I just had missed the review till today :)

WickedShell commented:
@OXINARF I've addressed your comments, and also attached one new commit that creates a enum for _auto_config meanings, because I started to feel nasty about hardcoding the meanings of 0/1 everywhere, and just treating it as a bool was actually not what we wanted to do...

WickedShell commented:
Rebased on master to address the MAVLink branches being deleted which was breaking CI.

OXINARF commented:
I've squashed @priseborough's commits, fixed a bug in the commit *AP_GPS: Allow backends to report their lag confidences* (which was corrected in the final commit) and pushed to a personal branch: https://github.com/ArduPilot/ardupilot/compare/master...OXINARF:review/6315
Final diff is exactly the same so I'm confident with that. Can it be merged? @priseborough, @WickedShell? This had an important bug fix after Paul tested, so it was probably good to test it again.

WickedShell commented:
I haven't tested it, the changes to my commits look reasonable though. @priseborough if you can test when you get a chance on @OXINARF's proposed set that would be great!

priseborough commented:
Test results with a UBlox M8P connected to GPS1 - see below for details

GPS_TYPE = 1, GPS_DELAY_MS = 0 (these are the parameter defaults)

**PASS**: It didn't wait for the GPS type to be determined and the EKF buffer was set for a GPS delay of 220msec

GPS_TYPE = 0, GPS_DELAY_MS = 0

**PASS**: EKF3 sets buffers immediately assuming a GPS lag of 0 (buffer length is determined by other sensor delays)

GPS_TYPE = 2, GPS_DELAY_MS = 0

**PASS**: EKF3 waits for configuration and sets buffers assuming a GPS lag of 120.

GPS_TYPE = 1, GPS_DELAY_MS = 100

**PASS**: EKF3 waits for configuration and sets buffers assuming a GPS lag of 100.

GPS_TYPE = 0, GPS_DELAY = 200

**PASS**: EKF3 sets buffers immediately assuming a GPS lag of 200.

OXINARF commented:
> GPS_TYPE = 1, GPS_DELAY_MS = 100
>
>PASS: EKF3 waits for configuration and sets buffers assuming a GPS lag of 100.

@priseborough Is this correct? In this case it shouldn't wait but immediately set buffers assuming a GPS lag of 100.

WickedShell commented:
I pinged him already on the same concern :)

On Tue, Jun 20, 2017 at 6:50 PM, Francisco Ferreira <
notifications@github.com> wrote:

> GPS_TYPE = 1, GPS_DELAY_MS = 100
>
> PASS: EKF3 waits for configuration and sets buffers assuming a GPS lag of
> 100.
>
> @priseborough <https://github.com/priseborough> Is this correct? In this
> case it shouldn't wait but immediately set buffers assuming a GPS lag of
> 100.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/6315#issuecomment-309939680>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAipiFDe7IvVNBqT4XrvsGT0x7GVmqZhks5sGHb_gaJpZM4NleEO>
> .
>

priseborough commented:
Sorry about that, I misread my testing notes. I've corrected it above.

OXINARF commented:
Merged, thanks!
